source/util/move_to_front.h

```diff
@@ -0,0 +1,318 @@
+  inline uint16_t& MutableRightOf(uint16_t handle) {
+    assert(handle);
+    return nodes_.at(handle).right;
+  }
+
+  // Returns mutable handle to parent of the node referenced by |handle|.
+  inline uint16_t& MutableParentOf(uint16_t handle) {
+    assert(handle);
+    return nodes_.at(handle).parent;
+  }
+
+  // Returns mutable timestamp of the node referenced by |handle|.
+  inline uint32_t& MutableTimestampOf(uint16_t handle) {
+    assert(handle);
+    return nodes_.at(handle).timestamp;
+  }
+
+  // Returns mutable size of the node referenced by |handle|.
+  inline uint16_t& MutableSizeOf(uint16_t handle) {
+    assert(handle);
+    return nodes_.at(handle).size;
+  }
+
+  // Returns mutable height of the node referenced by |handle|.
+  inline uint8_t& MutableHeightOf(uint16_t handle) {
+    assert(handle);
+    return nodes_.at(handle).height;
+  }
+
+  // Returns true iff |handle| is left child of its parent.
+  inline bool IsLeftChild(uint16_t handle) const {
+    assert(handle);
+    return LeftOf(ParentOf(handle)) == handle;
+  }
+
+  // Returns true iff |handle| is right child of its parent.
+  inline bool IsRightChild(uint16_t handle) const {
+    assert(handle);
+    return RightOf(ParentOf(handle)) == handle;
+  }
+
+  // Returns the sibling of handle or 0 if no sibling.
+  inline uint16_t SiblingOf(uint16_t handle) const {
+    assert(handle);
+    const uint16_t parent = ParentOf(handle);
+    return LeftOf(parent) == handle ? RightOf(parent) : LeftOf(parent);
+  }
+
+  // Returns the height difference between right and left subtrees.
+  inline int BalanceOf(uint16_t handle) const {
+    return int(HeightOf(RightOf(handle))) - int(HeightOf(LeftOf(handle)));
+  }
+
+  // Updates size and height of the node, assuming that the children have
+  // correct values.
+  void UpdateNode(uint16_t handle);
+
+  // Returns the most LeftOf(LeftOf(... descendent which is not leaf.
+  uint16_t LeftestDescendantOf(uint16_t handle) const {
+    uint16_t parent = 0;
+    while (handle) {
+      parent = handle;
+      handle = LeftOf(handle);
+    }
+    return parent;
+  }
+
+  // Returns the most RightOf(RightOf(... descendent which is not leaf.
+  uint16_t RightestDescendantOf(uint16_t handle) const {
+    uint16_t parent = 0;
+    while (handle) {
+      parent = handle;
+      handle = RightOf(handle);
+    }
+    return parent;
+  }
+
+  // Prints the internal tree data structure for debug purposes in the following
+  // format:
+  // 10H3S4----5H1S1-----D2
+  //           15H2S2----12H1S1----D3
+  // Right links are horisontal, left links step down one line.
+  // 5H1S1 is read as id 5, height 1, size 1. Optionally node label can also
+  // contain timestamp (5H1S1T15). D3 stands for depth 3.
+  void PrintTreeInternal(std::ostream& out, uint16_t node, size_t depth,
+                         bool print_timestamp) const;
+
+  // Inserts node handle in the tree.
+  void InsertNode(uint16_t node);
```

Comment:
I probably should replace "handle" with "node" in the accessor functions for consistency.